README.md
## Share-Primus

[Primus](https://github.com/primus/primus) bindings for [ShareJS](https://github.com/share/ShareJS).

Share-Primus is a client (browser side) extension to ShareJS that lets you use ShareJS with WebSockets,
BrowserChannel, SockJS or any other streaming library supported by Primus.

Primus also has several useful plugins such as [Substream](https://github.com/primus/substream), which
makes it possible to use the same connection as a transport for several logical streams.

### Browser

Import the scripts:

```html
<script src="/primus/primus.js"></script>
<script src="text.js"></script>
<script src="share.uncompressed.js"></script>
<script src="share-primus.js"></script>
```

Create a connection and subscribe to a document:

```javascript
var primus = new Primus();

var sjs = new window.sharejs.Connection(primus);
var doc = sjs.get('test-collection', 'test-doc');
doc.subscribe();

doc.whenReady(function () {
  if (!doc.type) doc.create('text');
  doc.attachTextarea(document.getElementById('pad'));
});
```

### Server (Node.js)

```javascript
var primus = new Primus(server);
var shareClient = share.server.createClient({backend:backend});

primus.on('connection', function (spark) {
  shareClient.listen(new sharePrimus.SparkStream(spark));
});
```

### Run Example

Start the server:

```
node example/server.js --transformer=websockets
```

You can also start the server with other streaming frameworks:

```
node example/server.js --transformer=[websockets|browserchannel|sockjs|engine.io|socket.io]
```

IMPORTANT! When you start the server the first time, Primus will tell you to `npm install`
the underlying streaming framework first. Pay close attention to the error message.

The example code is a little more sophisticated than the code above.
It also sets up a substream for sending non-sharejs messages over the same connection.

Open a browser: http://localhost:7008/


History.md
## [0.1.3](https://github.com/aslakhellesoy/share-primus/compare/v0.1.2...v0.1.3) (2014-11-26)

* Update dependencies
* Use in-memory livedb for example

## [0.1.2](https://github.com/aslakhellesoy/share-primus/compare/v0.1.1...v0.1.2) (2014-06-16)

* Work around ShareJS' strict state machine. Primus sometimes goes from
  `CLOSED` to `OPEN` without going via `OPENING` (Julien Biezemans, Aslak Hellesøy)

## [0.1.1](https://github.com/aslakhellesoy/share-primus/compare/v0.1.0...v0.1.1)

* Better error handling (Aslak Hellesøy)

## [0.1.0](https://github.com/aslakhellesoy/share-primus/compare/v0.0.3...v0.1.0) (2014-05-30)

* Cleanup spark_stream and add tests. ([#6](https://github.com/aslakhellesoy/share-primus/pull/6) Friedel Ziegelmayer)

## [0.0.3] (2014-02-23)

* First release


test/spark_stream.spec.js
var chai = require('chai')
chai.use(require('sinon-chai'));
var expect = chai.expect;
var EventEmitter = require('events').EventEmitter;
var sinon = require('sinon');
var Primus = require('primus');
var Spark = Primus.Spark;
var http = require('http');

var SparkStream = require('../lib/spark_stream');


describe('SparkStream', function () {
  it('is instantiable', function () {
    var spark = new Spark(new Primus(http.createServer()));
    expect(new SparkStream(spark)).to.be.an.instanceOf(SparkStream);
  });

  it('writes to the underlying spark', function () {
    var spark = new EventEmitter();
    spark.write = sinon.stub();
    var stream = new SparkStream(spark);
    stream.write('hello world');
    expect(spark.write).to.have.been.calledWith('hello world');
  });

  it('emits data from the underlying spark', function (done) {
    var spark = new EventEmitter();
    var stream = new SparkStream(spark);
    stream.on('data', function (data) {
      expect(data).to.be.eql('hello world');
      done();
    });
    spark.emit('data', 'hello world');
  });

  it('emits an end event when the spark stream closes', function (done) {
    var spark = new EventEmitter();
    var stream = new SparkStream(spark);

    stream.on('end', function () {
      done();
    });
    spark.emit('end');
  });
  it('emits a finish event when the spark stream closes', function (done) {
    var spark = new EventEmitter();
    var stream = new SparkStream(spark);

    stream.on('finish', function () {
      done();
    });
    spark.emit('end');
  });
  it('emits an error when the spark stream has an error', function (done) {
    var spark = new EventEmitter();
    var stream = new SparkStream(spark);

    stream.on('error', function (err) {
      expect(err).to.be.eql('error');
      done();
    });
    spark.emit('error', 'error');
  });
});


lib/spark_stream.js
var Duplex = require('stream').Duplex;
var inherits = require('util').inherits;

// Workaround for https://github.com/primus/primus/issues/121


function SparkStream(spark) {
  if (!(this instanceof SparkStream)) return new SparkStream(spark);

  Duplex.call(this, {
    objectMode: true,
    allowHalfOpen: false
  });

  // Store a reference to the underlying spark stream
  this._spark = spark;

  // Listen to data events from the spark and push them
  // into the stream.
  spark.on('data', this.push.bind(this));

  // Listen to the close event of the spark stream.
  //spark.on('close', this.push.bind(this, null));
  spark.on('end', function () {
    this.resume();
    this.push(null);
  }.bind(this));

  // Emit error events
  spark.on('error', this.emit.bind(this, 'error'));
}

inherits(SparkStream, Duplex);

// Implement the `_write` method for the writable part of the
// Duplex stream.
SparkStream.prototype._write = function (chunk, encoding, cb) {
  this._spark.write(chunk);
  cb();
};

// Implement the `_read` method for th readable part of the
// Duplex stream.
SparkStream.prototype._read = function () {
};


module.exports = SparkStream;


lib/client/share-primus.js
(function (window) {
  'use strict';

  // Map Primus ready states to ShareJS ready states.
  var STATES = {};
  STATES[Primus.CLOSED] = 'disconnected';
  STATES[Primus.OPENING] = 'connecting';
  STATES[Primus.OPEN] = 'connected';

  // Override Connection's bindToSocket method with an implementation
  // that understands Primus Stream.
  window.sharejs.Connection.prototype.bindToSocket = function(stream) {
    var connection = this;

    setState(stream.readyState);
    this.canSend = this.state === 'connected'; // Primus can't send in connecting state.

    // Tiny facade so Connection can still send() messages.
    this.socket = {
      send: function(msg) {
        try {
          stream.write(msg);
        } catch(e) {
          connection.emit('error', e);
        }
      }
    };

    stream.on('data', function(msg) {
      try {
        connection.handleMessage(msg);
      } catch (e) {
        connection.emit('error', e);
      }
    });

    stream.on('readyStateChange', function() {
      setState(stream.readyState);
    });

    function setState(readyState) {
      var shareState = STATES[readyState];
      if(readyState === Primus.OPEN && connection.state === STATES[Primus.CLOSED]) {
        // Hack to work around ShareJS' strict state machine.
        // Sometimes Primus seems to go straight from CLOSED to OPEN without an OPENING in-between.
        connection._setState(STATES[Primus.OPENING]);
      }
      connection._setState(shareState);
    }
  };
})(window);


example/server.js
var format = require('util').format;
var http = require('http');
var express = require('express');
var livedb = require('livedb');
var share = require('share');
var sharePrimus = require('../lib');
var Primus = require('primus');
var argv = require('optimist').argv;

var app = express();
app.use(express.static(__dirname));
app.use(express.static(share.scriptsDir));
app.use(express.static(sharePrimus.scriptsDir));
var server = http.createServer(app);

var primus = new Primus(server, { transformer: argv.transformer });
primus.use('substream', require('substream'));

var db = new livedb.memory();
var backend = livedb.client(db);
var shareClient = share.server.createClient({backend:backend});

primus.on('connection', function (spark) {
  console.log('New client');
  var shareSpark = spark.substream('share');
  // Workaround for https://github.com/primus/primus/issues/121
  shareClient.listen(new sharePrimus.SparkStream(shareSpark));

  // Send some non-share messages over the same stream
  // Inspired by https://github.com/einaros/ws/blob/master/examples/serverstats/server.js
  var processSpark = spark.substream('process');
  var id = setInterval(function() {
    processSpark.write(JSON.stringify(process.memoryUsage()));
  }, 100);

  spark.on('end', function() {
  	console.log('Lost client');
  	clearInterval(id);
  });
});

var port = argv.port || 7008;
server.listen(port);
console.log(format("Listening on http://localhost:%d/", port));
